src/jaxsim/integrators/variable_step.py

```diff
@@ -0,0 +1,610 @@
+        # We run these loops until the final time tf is reached.
+        def while_loop_body(carry: Carry) -> Carry:
+
+            # Unpack the carry.
+            x0, t0, params, discarded_steps, _ = carry
+
+            # Take care of the final adaptive step.
+            # We want the final Δt to let us reach tf exactly.
+            # Then we can exit the while loop.
+            Δt0 = params["dt0"]
+            Δt0 = jnp.where(t0 + Δt0 < tf, Δt0, tf - t0)
+            break_loop = jnp.where(t0 + Δt0 < tf, False, True)
+
+            # Run the underlying explicit RK integrator.
+            # The output z contains multiple solutions (depending on the rows of b.T).
+            with self.editable(validate=True) as integrator:
+                integrator.params = params
+                z = integrator._compute_next_state(x0=x0, t0=t0, dt=Δt0, **kwargs)
+                params_next = integrator.params
+
+            # Extract the high-order solution xf and the low-order estimate x̂f.
+            xf = jax.tree_util.tree_map(lambda l: l[self.row_index_of_solution], z)
+            x̂f = jax.tree_util.tree_map(
+                lambda l: l[self.row_index_of_solution_estimate], z
+            )
+
+            # Calculate the local integration error.
+            local_error = local_error_estimation(
+                x0=x0, xf=xf, xf_estimate=x̂f, rtol=self.rtol, atol=self.atol
+            )
+
+            # Shrink the Δt every time by the safety factor (even when accepted).
+            # The β parameters define the bounds of the timestep update factor.
+            safety = jnp.clip(self.safety, a_min=0.0, a_max=1.0)
+            β_min = jnp.maximum(0.0, self.beta_min)
+            β_max = jnp.maximum(β_min, self.beta_max)
+
+            # Compute the next Δt from the desired integration error.
+            # The computed integration step is accepted if error <= 1.0,
+            # otherwise it is rejected.
+            #
+            # In case of rejection, Δt_next is always smaller than Δt0.
+            # In case of acceptance, Δt_next could either be larger than Δt0,
+            # or slightly smaller than Δt0 depending on the safety factor.
+            Δt_next = Δt0 * jnp.clip(
+                a=safety * jnp.power(1 / local_error, 1 / (q + 1)),
+                a_min=β_min,
+                a_max=β_max,
+            )
+
+            def accept_step():
+                # Use Δt_next in the next while loop.
+                # If it is the last one, and Δt0 was clipped, return the initial Δt0.
+                params_next_accepted = params_next | dict(
+                    dt0=jnp.clip(
+                        jax.lax.select(
+                            pred=break_loop,
+                            on_true=params["dt0"],
+                            on_false=Δt_next,
+                        ),
+                        self.dt_min,
+                        self.dt_max,
+                    )
+                )
```

Comment:
Calls to `dict` [have been shown](https://madebyme.today/blog/python-dict-vs-curly-brackets/) to be slower than just using the dict literal. We might prefer to use those

```suggestion
                params_next_accepted = params_next | {
                    "dt0": jnp.clip(
                        jax.lax.select(
                            pred=break_loop,
                            on_true=params["dt0"],
                            on_false=Δt_next,
                        ),
                        self.dt_min,
                        self.dt_max,
                    )
                }
```

Reply:
Remember that this is true only in plain Python, not in jit-compiled code.